### PR TITLE
Alternative Regular Expressions for Scala-Native

### PR DESCRIPTION
**Pull request by muxanick**
These are alternative classes for regular expressions built on top of c++11 regex code.
It supports missing features like back tracing and requires no additional libraries on the machine. Full support for Unicode etc.
It's not replacement for re2 code but co-existing in nativelib project for mostly internal usage, but nothing stops you to use it from another project.
It allows us to simplify many algorithms like url parser, ip address parser, formatter etc. It also allows to reuse regex from scala.js because of high compatibility with JavaScript regex (I would say 99.9% compatible).
Primary goals for code were runtime performance and scala-style interfaces, including `Scope.scala` created by @densh which allows us to use memory more effective without zombi allocations.

Please check `RegexSuite.scala` for examples.

### CODE REVIEW

**Comment by muxanick** [comment]
Travis said that I haven't signed CLA, but I did it before, I don't know why it asks again, so I signed CLA again.

**Comment by muxanick** [comment]
CI breaks are not related to the change

**Comment by densh** [comment]
@muxanick Thanks a lot for the suggestion but I'm afraid we can't maintain more than one implementation of regular expressions in the core. For C++11 regular expressions to be accepted they have to replace RE2 completely.

**Comment by densh** [comment]
Two options we have going forward with this PR:

1. Make it a stand-alone library.

2. Make C++11-based regex supersede RE2-based ones by replacing the implementation of `java.util.regex.*`.

File: bin/travis_setup.sh [hunk 1]
```diff
@@ -22,9 +22,11 @@ else
     sudo add-apt-repository --yes ppa:ubuntu-toolchain-r/test
     sudo apt-get -qq update
     sudo apt-get install -y -qq \
+      gcc-5 g++-5 \
```
Comment (densh) [inline bin/travis_setup.sh]: Adding more dependencies to the core is really hard to justify. Every single additional dependency makes it harder to setup environment on older Linux distributions.

File: nativelib/src/main/scala/scala/scalanative/misc/CRegex.scala [hunk 2]
```diff
@@ -0,0 +1,106 @@
+package scala.scalanative.misc
+package regex
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/misc/CRegex.scala]: We've been struggling really hard not to introduce new APIs outside in core with the exception of those necessary for interop. The reason being is that as soon as you use any of the non-standard APIs your code becomes completely unportable wrt JVM and JS targets.

File: nativelib/src/main/scala/scala/scalanative/misc/Scope.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+@annotation.implicitNotFound(msg = "Resource acquisition requires a scope.")
+trait Scope {
+
+  /** Push resource onto the resource stack. */
+  def acquire[R <: Resource](res: R): R
+
+  /** Clean up all the resources in FIFO order. */
+  def close(): Unit
+}
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/misc/Scope.scala]: These two are extremely valuable and deserve a separate PR. Given that scope/resource API is platform independent we can cross-publish it to be available on JVM and JS.

Comment (muxanick) [inline nativelib/src/main/scala/scala/scalanative/misc/Scope.scala]: I'm just copied it from tools, it's your file :)

Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/misc/Scope.scala]: Sure! This one deserves it's own module which is cross-published for JVM and JS. Tools can just depend on it.

File: bin/travis_setup.sh (hunk 1 above)
Comment (muxanick) [inline bin/travis_setup.sh]: It is only for linux plarform.
GCC 4.9.0,	April 22, 2014
I agreed with you that it's very risky.
I found that scala-native doesn't work with java 1.7, it requires to upgrade to java 1.8 which is not supported on Ubuntu 14.04 for instance (there is third party package but it's not stable, shows errors)
I wasn't expected you accept this PR :) I just can't work on separate repository until scala-native becomes more mature.

File: nativelib/src/main/scala/scala/scalanative/misc/Scope.scala (hunk 3 above)
Comment (muxanick) [inline nativelib/src/main/scala/scala/scalanative/misc/Scope.scala]: I would not make this module myself, it seems not right to make a module from file which is written by other person.

Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/misc/Scope.scala]: I mean a separate subproject in our sbt build inside of the main repo. It's perfectly fine to move stuff around as part of a PR. :)

Comment (muxanick) [inline nativelib/src/main/scala/scala/scalanative/misc/Scope.scala]: When you said `These two are`, which is second?

Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/misc/Scope.scala]: Scope and Resource. Btw Resource can just be an alias to AutoCloseable.

Comment (muxanick) [inline nativelib/src/main/scala/scala/scalanative/misc/Scope.scala]: AutoCloseable crashes the compiler, pretty much anything from other libs is crashing the compiler

Comment (xuwei-k) [inline nativelib/src/main/scala/scala/scalanative/misc/Scope.scala]: > AutoCloseable crashes the compiler

https://github.com/scala-native/scala-native/pull/853 fixed?

Comment (muxanick) [inline nativelib/src/main/scala/scala/scalanative/misc/Scope.scala]: I will check, thanks
